### DISCUSSION TITLE
Support Tool Calls / Functions in Playground

### DISCUSSION THREAD

**Discussion by yuyuma** [langfuse/langfuse#3166] [Ideas]
### Describe the feature or potential improvement

Currently the playground only supports System, User, and Assistant messages.  Requesting support for tool calls so that the playground can be used to iterate on function calling prompts.

This request extends to testing existing tool use generation traces in the playground.

**Comment by marcklingen** [comment]
thanks for sharing!

**Comment by ghost** [comment]
Bumping this since it prevents us from using the playground effectively.

**Comment by hassiebp** [comment]
Related issue: https://github.com/langfuse/langfuse/issues/3931

**Comment by ajram23** [comment]
+1 to this.

**Comment by dquang-tran** [comment]
+1 on not being able to use playground without this

**Comment by sang4lv** [comment]
+1 tool call testing is badly in need 🙋🏻‍♀️

**Comment by jeff-an** [comment]
+1 to this, playground is basically unusable without it. other llm observability providers have this already

**Comment by cosasha97** [comment]
+1 to this, I agree with @jeff-an !

**Comment by ClementNguyen** [comment]
+1!!

**Comment by DovahkiinM** [comment]
+1, really need this to use playground

**Comment by david1542** [comment]
+1, need this as well! 🚀

**Comment by theo-michel** [comment]
+1 this should be top priority, most of llm use is structured output today, and it's not that hard to add

**Comment by Pringled** [comment]
+1, this would be very useful. Structured outputs are a must for testing & developing prompts in the playground.

**Comment by emieldatalytica** [comment]
I understood from the Langfuse team that supporting tool calls in the playground is on the immediate roadmap. 

However, that feature (supporting structured outputs) will currently only be extended to the playground. In other words, it won't enable doing prompt experiments from within the UI on a dataset, while applying tool calls or structured outputs. There is a separate discussion opened for this, see [here](https://github.com/orgs/langfuse/discussions/5958) if you'd like to mix in the discussion.

**Comment by hassiebp** [comment]
We have released tool call and structured output support for playground 🥳 

https://langfuse.com/changelog/2025-03-28-tool-calling-structured-output-playground

**Reply by jeff-an** [reply]
Thanks for the update @hassiebp ! as a fast follow would it be possible to auto-import the tools from my LLM call when I go from trace -> playground? otherwise, re-creating the exact tool definition can be cumbersome for each trace I inspect and experiment with.

**Reply by ajram23** [reply]
+1 to this please.

**Reply by hassiebp** [reply]
Yes, we have follow up tasks to write tool definition / call parsers from the trace format of our various integrations to the playground. Next up is Langchain. @jeff-an could you share a trace link or the integration from which you capture tool call traces?

**Reply by ajram23** [reply]
@hassiebp here is mine with langchain & tool calling. You know how to reach me if you need more info. https://us.cloud.langfuse.com/project/cltolaqjv0000m2pkhvdzv9f6/traces/c212e2bd-4774-4511-b2e1-b128e0b1fa2c?timestamp=2025-03-31T02:26:26.500Z&display=details

**Reply by hassiebp** [reply]
Thanks AJ!

**Reply by hassiebp** [reply]
@ajram23 @jeff-an Automatic tool parsing from Langchain generations when jumping into the playground is now live - please try it out and let me know your feedback! 🙏🏾

**Reply by ajram23** [reply]
@hassiebp very cool and very well done 👏🏼 cc @marcklingen - Hassieb rocks!!. 3 questions 1. Just like prompts you already know the various tool calls from traces (I wish you differentiate it in the trace UI instead of a span), why does it have to saved aka why the unsaved status? 2. Very very interesting to see Structured Outputs - is the plan to support pydantic classes/output? 3. Can you kindly also import the input message from the trace?

**Reply by pokliu** [reply]
```
_app-d5d8f035227b6c53.js:30 TypeError: Cannot read properties of undefined (reading 'length')
at playground-f860159be81f4acd.js:1:7824
at Array.filter (<anonymous>)
at playground-f860159be81f4acd.js:1:7804
at playground-f860159be81f4acd.js:1:8003
at onClick (playground-f860159be81f4acd.js:1:21513)
at Object.eB (framework-c511aa5252cf0ba6.js:1:84184)
at eH (framework-c511aa5252cf0ba6.js:1:84338)
at framework-c511aa5252cf0ba6.js:1:102644
at re (framework-c511aa5252cf0ba6.js:1:102743)
at rt (framework-c511aa5252cf0ba6.js:1:103157)
at framework-c511aa5252cf0ba6.js:1:108613
at uP (framework-c511aa5252cf0ba6.js:1:165580)
at eM (framework-c511aa5252cf0ba6.js:1:83322)
at ru (framework-c511aa5252cf0ba6.js:1:104448)
at tB (framework-c511aa5252cf0ba6.js:1:90117)
at tD (framework-c511aa5252cf0ba6.js:1:89901)
at HTMLDivElement.n (_app-d5d8f035227b6c53.js:27:4756)
```

``` javascript
 return t.filter(e => e.content.length > 0 || "toolCalls"in e && e.toolCalls.length > 0)
```

When jumping from the trace generation to the playground, my `toolcall` `assistant` message has no `content` field, which causes a JS error and an alert popup when clicking submit button. I think no `content` field should be supported.

**Reply by hassiebp** [reply]
@pokliu Thanks for raising this - could you please provide a link to the generation from which you jump to the playground here in this issue? https://github.com/langfuse/langfuse/issues/6324

@ajram23 Thanks a lot for your feedback! 🙏🏾🫰🏾

> 1. Just like prompts you already know the various tool calls from traces (I wish you differentiate it in the trace UI instead of a span), why does it have to saved aka why the unsaved status?

We decided to not automatically save every tool definition ever used in a generation to a user's project. Once saved, it should show as saved every time you jump into the playground with a tool definition that has been previously saved (same name, description, and schema)

> 2. Very very interesting to see Structured Outputs - is the plan to support pydantic classes/output?

We are currently supporting raw JSONSchema only and do not have immediate plans for Pydantic support (Python) or zod schema support (JS/TS). You can however generate a jsonschema via this Pydantic utility method: https://docs.pydantic.dev/latest/concepts/json_schema/ - similar functionality is used by Langchain / Openai SDKs to convert Pydantic / zod schemas to JSONSchema that is the format accepted by the OpenAI API

> 3. Can you kindly also import the input message from the trace?

Could you please explain what you mean exactly / share a trace link from which this issue is apparent?

**Reply by ajram23** [reply]
@hassiebp 1. Ok, but at the very least can you please distinguish between tools and spans in the detailed view.  Hard to debug other wise. 
2.sure, can do. But the larger question is why just support structured outputs just in the playground? 
3. Oh I just meant the input text that was passed to it from either a prior block or the user input. 

Thanks.

**Comment by marctorsoc** [comment]
@hassiebp here is a [generation](https://us.cloud.langfuse.com/project/cmd300b5k01pzad07ds1w97a3/traces?peek=3fcb96ce996e07837a6b2b5f26a88204&timestamp=2025-07-18T10%3A26%3A17.506Z&observation=496207d36cd193ac) that does not allow me to jump to the playground, if that helps. And I copied the input to the LLM (I had to anonymize it, hope it still fails if someone attempts to try it)
```
[
    {
        "role": "system",
        "content": "\n    <base_instructions>\n    You are Ana, a professional and friendly agent. You work for XX</base_instructions><agent_instructions>\n    Now ask the user if they want to hear about the highlights of the contract or they prefer to \n    read it on the website. If they don't, then ask if they have any questions. \n\n    Use the <contract_highlights> section from your context for the highlights. \n    Please explain everything in a friendly and professional manner, but without being too boring/technical.\n\n    Then, use everything in your context to answer their questions. Please do not make up anything. Skip \n    the contract terms unless you cannot find the information elsewhere, \n    and if they ask a question present in the FAQs, answer **exactly** as it is written therein. \n    If you don't have the information, just say that you don't know, but they can check the website for more details,\n    or you can flag this to the facility manager who will call them in the next 24 hours to answer their questions.\n\n    Once they don't have any more questions, call the tool to transfer to the \n    next agent, which is `extend_contact_details`.\n\n    If they decide to correct the selected unit, use the tool to go back to re-select the unit.\n    </agent_instructions>\n    <context><time_now>```Friday, 2025-07-18 12:30:41```</time_now><contract_highlights>You need to stay with us forever and pay 10% more every year</contract_highlights></context>"
    },
    {
        "role": "assistant",
        "content": "Okay, so you're looking to start this Sunday, July 20th, 2025. Got it.\n\nBased on ...",
    },
    {"role": "user", "content": "I think i'm fine with ..."},
    {
        "role": "assistant",
        "content": "I understand you're looking for ... Unfortunately, the ... is currently unavailable.\n\I would still recommend the ... for ... dollars per month. Would you like to proceed with the ...?",
    },
    {"role": "user", "content": "ok"},
    {
        "role": "assistant",
        "content": "Great! The ... sounds like a perfect fit for you.\n\nAre you ok to confirm?",
    },
    {"role": "user", "content": "sure"},
    {
        "role": "assistant",
        "tool_calls": [
            {
                "function": {"name": "collect_selected_unit", "arguments": {"model_id": "75200"}},
                "id": "function_call_4bb55e48c90d",
                "type": "function",
            }
        ],
    },
    {
        "role": "tool",
        "content": "The selected unit is `75200`",
        "name": "collect_selected_unit",
        "id": "function_call_4bb55e48c90d",
    },
]
```
Please let me know if I can help you fix the issue somehow, or I'm doing something wrong. This ended up in langfuse by using the recent (yesterday was released) livekit-agents integration

**Reply by hassiebp** [reply]
Could you please share a link to the trace? Or the project ID / trace ID from the URL?

**Reply by marctorsoc** [reply]
here it is @hassiebp https://us.cloud.langfuse.com/project/cmd300b5k01pzad07ds1w97a3/traces/9d7a4ad290b4f8c4834ae34f27b1d9c2?timestamp=2025-07-18T11%3A15%3A37.762Z&display=details&observation=f4da73da7a745425

**Reply by marctorsoc** [reply]
ok, I think I got the issue. I cannot jump to playground for llm_requests where there is the output of a tool call. Is this right? The problem  is I have a multi-agent system, and each agent asks a number of questions. I can only jump to playground before they store (with a tool) one of the slots they are responsible of. After that, all the other turns cannot be simulated in the playground, which makes this feature way less useful than could be. Is there some way I could fix this? Maybe logging differently? Or something you could do on your end?

**Comment by mrdrprofuroboros** [comment]
@hassiebp can we import tool definitions from the trace that was used to populate the playground?

we have tools that get one of the enum parameters being dynamically filtered in runtime depending on the generation. so I can't define a single reusable tool, i want to populate it from the historical generation with the exactly same enum options that were used there so I'd be able to tweak prompts in these conditions
